### ISSUE TITLE
Federation library failing to compile with tsc

### ISSUE THREAD

**Issue by zackpollard** [apollographql/federation#1538]
## Last working version
@apollo/federation@0.33.9

## Broken versions
Anything newer than @apollo/federation@0.33.9

## Expected Behaviour
Apollo federation should compile with `tsc -p tsconfig.json --noEmit`

## Actual Behaviour
Apollo federation fails to compile due to composition/utils.d.ts referencing a namespace that the compiler cannot find.
```
node_modules/@apollo/federation/dist/composition/utils.d.ts:43:371 - error TS2503: Cannot find namespace 'Kind'.

43     kind: (Kind.SCALAR_TYPE_DEFINITION | Kind.OBJECT_TYPE_DEFINITION | Kind.INTERFACE_TYPE_DEFINITION | Kind.UNION_TYPE_DEFINITION | Kind.ENUM_TYPE_DEFINITION | Kind.INPUT_OBJECT_TYPE_DEFINITION | Kind.DIRECTIVE_DEFINITION | Kind.SCALAR_TYPE_EXTENSION | Kind.OBJECT_TYPE_EXTENSION | Kind.INTERFACE_TYPE_EXTENSION | Kind.UNION_TYPE_EXTENSION | Kind.ENUM_TYPE_EXTENSION | Kind.INPUT_OBJECT_TYPE_EXTENSION)[];
```

I don't have the time right now to put together a reproduction repo, but if this is required I can look into doing it at a later point. I've seen people mentioning this same issue on stack overflow but can't see an issue anywhere in the issue tracker.

## Typescript Config
I've added the typescript config that we use as that's the main thing I can think of that might influence this to fail.
```
{
  "compilerOptions": {
    "target": "es2017",
    "module": "commonjs",
    "strict": true,
    "baseUrl": ".",
    "paths": {
      "*": ["src/*"],
    },
    "typeRoots": ["node_modules/@types"],
    "types": ["node", "jest"],
    "esModuleInterop": true,
    "inlineSourceMap": true,
    "allowJs": false,
    "resolveJsonModule": true,
    "strictFunctionTypes": false,
  }
}
```

**Comment by cpeacock** [comment]
Thanks for submitting the issue!  We'll take a look at this.

**Comment by clenfest** [comment]
Hey @chrskrchr, thanks for pushing on this. I've created a PR for this issue, but I'd like to make sure that it's well reviewed before pushing a release. Shouldn't be too long.

**Comment by std4453** [comment]
#1585 hasn't solved this issue.

In `@apollo/federation@0.35.3`, `diffTypeNodes().type` has type `(Kind.SCALAR_TYPE_DEFINITION | Kind.OBJECT_TYPE_DEFINITION | ... )[]`.

After #1585, in `@apollo/federation@0.35.4`, `kind` field of `diffTypeNodes` is `Kind[]`, however that is still not compatible with `graphql@15.8.0` where `Kind` is a `const` and cannot be used as type.

So now the error message has changed to:

```
../../common/temp/node_modules/.pnpm/@apollo+federation@0.35.4_graphql@15.8.0/node_modules/@apollo/federation/dist/composition/utils.d.ts(43,11): error TS2749: 'Kind' refers to a value, but is being used as a type here. Did you mean 'typeof Kind'?
```

**Comment by std4453** [comment]
Looks like that only way to keep compatible with both `graphql@15` and `graphql@16` is to use `KindEnum`, which is currently deprecated and will be removed in v17.

Otherwise, `@apollo/federation` and `@apollo/gateway` should not claim to be compatible with `graphql@15`.

**Comment by clenfest** [comment]
Ok, thanks for following up on this. I was trying to avoid using `KindEnum` since it's deprecated. Apparently I was leaning too heavily into my repro of the bug which obviously didn't cover your case sufficiently, though it fixed the original error for me. I'll take another look at this shortly.

**Comment by clenfest** [comment]
Oh interesting. I tried setting the type to `typeof Kind[keyof typeof Kind][]`, but I didn't realize that `tsc` would translate that as `Kind[]` in the `.d.ts` file, I thought it was pass it through as I'd written it.

**Comment by chrskrchr** [comment]
Confirmed that we're now running into the same build error as @std4453 in this comment:

https://github.com/apollographql/federation/issues/1538#issuecomment-1073205865

FWIW - we're open to migrating to `graphql@16` if that's a more reasonable path forward.  We'd just need to work past the peer dependency warnings and startup issue described in this comment:

https://github.com/apollographql/federation/issues/1538#issuecomment-1064420625

**Comment by dchambers** [comment]
Reproduction here in case it helps:

- https://stackblitz.com/edit/node-7sp9dl?file=index.ts

**Comment by chrskrchr** [comment]
FYI - @clenfest's [federation#1636](https://github.com/apollographql/federation/pull/1636) PR addressed this for us in  `@apollo/gateway@0.50.0`.

**Comment by trevor-scheer** [comment]
Thanks for the follow up @chrskrchr! I'll close this out.
